File: misc/tournament.js
// Generate a tournament schedule for n teams, subject to the following
// rules:
// 1. Each team has to play every other team.
// 2. Each team plays at most once a day.
//
// Questions:
// 1. How many days does the tournament last?
// 2. How many matches are there each day?

function generateSchedule(teams) {
    // Print the tournament schedule for the specified 'teams' to stdout.

    let size = teams.length;
    let days = size - 1;
    let matches = size / 2;
    for (let day = 0; day < days; ++day) {
        console.log("Day", day + 1);
        for (let i = 0; i < matches; ++i) {
            let a = teams[i];
            let b = teams[i + matches];
            console.log("    Team " + a + " vs. Team " + b);
        }
        rotate(teams);
    }
}

function rotate(teams) {
    // Rotate the specified 'teams' in-place such that the first element is
    // unchanged, and the rest of the array is shifted left by one position.

    let size = teams.length;
    let temp = teams[1];
    for (let i = 1; i < size - 1; ++i) {
        teams[i] = teams[i + 1];
    }
    teams[size - 1] = temp;
}

function main() {
    // Generate a tournament schedule for teams of size 'n'.

    let n = 8;
    let teams = [];
    for (let i = 0; i < n; ++i) {
        teams.push(String.fromCharCode("A".charCodeAt(0) + i));
    }
    generateSchedule(teams);
}

main();

File: misc/tournament.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <math.h>
using namespace std;

// Given N teams, generate a tournament schedule such that each team
// plays every other team. In each round, each team can only play one other 
// team.
// For example, 4 teams: A B C D
// Schedule:
// Round 1: A vs B, C vs D
// Round 2: A vs C, B vs D
// Round 3: A vs D, B vs C
// There are always N - 1 rounds

void print(char A, char B, unsigned round) {
  cout << "Round " << round << ": " << A << " vs. " << B << endl;
}

void generate_schedule(vector<char> teams, unsigned round) {
  unsigned N = teams.size();
  if (N <= 1) return;
  if (N == 2) {
    print(teams[0], teams[1], round);
    return;
  }
  for (unsigned i = 0; i < N / 2; i++) {
    for (unsigned j = 0; j < ceil(0.5 * N); j++) {
      print(teams[j], teams[j + N / 2], round);
    }
    round++;
    auto mid = teams.begin() + N / 2;
    rotate(mid, mid + 1, teams.end());
  }
  generate_schedule(vector<char>(teams.begin(), teams.begin() + N / 2), round);
  generate_schedule(vector<char>(teams.begin() + N / 2, teams.end()), round);
}


// for now assume a power of 2
int main() {
  char teams[26];
  for (unsigned i = 0; i < 26; ++i) {
    teams[i] = (char)('A' + i);
  }
  while(1) {
    cout << "Enter number of teams: ";
    unsigned total;
    cin >> total;
    generate_schedule(vector<char>(teams, teams + total), 1);
  }
}
